app/post_persistent_notification.go

```diff
@@ -0,0 +1,380 @@
+
+	// fetch posts for which first notificationInterval duration has passed
+	maxCreateAt := time.Now().Add(-notificationInterval).UnixMilli()
+
+	pagination := model.CursorPagination{
+		Direction: "down",
+		PerPage:   500,
+	}
+
+	// Pagination loop
+	for {
+		notificationPosts, hasNext, err := a.Srv().Store().PostPersistentNotification().Get(model.GetPersistentNotificationsPostsParams{
+			MaxCreateAt:   maxCreateAt,
+			MaxLastSentAt: maxCreateAt,
+			Pagination:    pagination,
+		})
+		if err != nil {
+			return errors.Wrap(err, "failed to get posts for persistent notifications")
+		}
+
+		// No posts available at the moment for persistent notifications
+		if len(notificationPosts) == 0 {
+			return nil
+		}
+		pagination.FromID = notificationPosts[len(notificationPosts)-1].PostId
+		pagination.FromCreateAt = notificationPosts[len(notificationPosts)-1].CreateAt
+
+		notificationPostsMap := make(map[string]*model.PostPersistentNotifications, len(notificationPosts))
+		postIds := make([]string, 0, len(notificationPosts))
+		for _, p := range notificationPosts {
+			postIds = append(postIds, p.PostId)
+			notificationPostsMap[p.PostId] = p
+		}
+		posts, err := a.Srv().Store().Post().GetPostsByIds(postIds)
+		if err != nil {
+			return errors.Wrap(err, "failed to get posts by IDs")
+		}
+
+		var expiredPosts []*model.Post
+		var validPosts []*model.Post
+		for _, p := range posts {
+			if notificationPostsMap[p.Id].SentCount >= notificationMaxCount {
+				expiredPosts = append(expiredPosts, p)
+			} else {
+				validPosts = append(validPosts, p)
+			}
+		}
+
+		// Delete expired notifications posts
+		expiredPostsIds := make([]string, 0, len(expiredPosts))
+		for _, p := range expiredPosts {
+			expiredPostsIds = append(expiredPostsIds, p.Id)
+		}
+		if err := a.Srv().Store().PostPersistentNotification().Delete(expiredPostsIds); err != nil {
+			return errors.Wrapf(err, "failed to delete expired notifications: %v", expiredPostsIds)
+		}
+
+		// Send notifications to validPosts
+		if err := a.forEachPersistentNotificationPost(validPosts, a.sendPersistentNotifications); err != nil {
+			return err
+		}
+
+		// Update last activity for valid notifications posts
+		validPostsIds := make([]string, 0, len(validPosts))
+		for _, p := range validPosts {
+			validPostsIds = append(validPostsIds, p.Id)
+		}
+		if err := a.Srv().Store().PostPersistentNotification().UpdateLastActivity(validPostsIds); err != nil {
+			return errors.Wrapf(err, "failed to update lastSentAt for valid notifications: %v", validPostsIds)
+		}
+
+		// Break pagination loop
+		if !hasNext {
+			break
+		}
+	}
+	return nil
+}
+
+func (a *App) forEachPersistentNotificationPost(posts []*model.Post, fn func(post *model.Post, channel *model.Channel, team *model.Team, mentions *ExplicitMentions, profileMap model.UserMap, channelNotifyProps map[string]map[string]model.StringMap) error) error {
+	channelIds := make(model.StringSet)
```

Comment:
I would extract half of this function away to give it a name. For example, in this case, I would use something like `getPersistentNotificationsAuxiliaryData` and would return a struct containing the maps that you are generating there. This function is really hard to read now.

Reply:
> This function is really hard to read now.

Now, try reading the original `SendNotifications` function. 😛